src/Compilers/Core/AnalyzerDriver/AnalyzerExecutor.cs

```diff
@@ -466,7 +476,18 @@ internal static bool CanHaveExecutableCodeBlock(ISymbol symbol)
 
         internal void ExecuteAndCatchIfThrows(DiagnosticAnalyzer analyzer, Action analyze)
         {
-            ExecuteAndCatchIfThrows(analyzer, analyze, _onAnalyzerException, _cancellationToken);
+            object gate;
+            if (_singleThreadedAnalyzerToGateMap.TryGetValue(analyzer, out gate))
+            {
+                lock(gate)
```

Comment:
AsyncSemaphore has been eliminated since the Workspaces layer moved to .Net 4.5.2. We just use SemaphoreSlim now. Eventually, you'll be targeting CoreCLR at this layer, so you could use SemaphoreSlim too.